beamer/state_machine.py

```diff
@@ -61,6 +61,14 @@ class Context:
     l1_invalidations: dict[ClaimId, Future]
     finality_periods: dict[ChainId, int] = field(default_factory=dict)
 
+    @property
+    def source_rpc_url(self) -> URL:
+        return URL(self.request_manager.web3.provider.endpoint_uri)  # type: ignore
```

Comment:
Why not cast the provider to `HTTPProvider` and remove the `# type: ignore` thing?